File: Language/C#/Book/tmp/LINQ实战/Chapter12.CSharp/LinqToAmazon/AmazonBookQueryProvider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Text;

namespace LinqInAction.Chapter12.LinqToAmazon
{
  public class AmazonBookQueryProvider : QueryProvider
  {
    public override String GetQueryText(Expression expression)
    {
      AmazonBookQueryCriteria criteria;

      // Retrieve criteria
      criteria = new AmazonBookExpressionVisitor().ProcessExpression(expression);

      // Generate URL
      String url = AmazonHelper.BuildUrl(criteria);

      return url;
    }

    public override object Execute(Expression expression)
    {
      String url = GetQueryText(expression);
      IEnumerable<AmazonBook> results = AmazonHelper.PerformWebQuery(url);
      return results;
    }
  }
}

File: Language/C#/Book/tmp/LINQ实战/Chapter12.VB/LinqToAmazon/AmazonBookQueryProvider.vb
Imports System.Linq.Expressions

Public Class AmazonBookQueryProvider
  Inherits QueryProvider

  Public Overloads Overrides Function GetQueryText(ByVal expression As Expression) As String
    Dim criteria As AmazonBookQueryCriteria

    ' Retrieve criteria
    criteria = New AmazonBookExpressionVisitor().ProcessExpression(expression)

    ' Generate URL
    Dim url As String = AmazonHelper.BuildUrl(criteria)

    Return url
  End Function

  Public Overloads Overrides Function Execute(ByVal expression As Expression) As Object
    Dim url As String = GetQueryText(expression)
    Dim results As IEnumerable(Of AmazonBook) = AmazonHelper.PerformWebQuery(url)
    Return results
  End Function
End Class